File: opengrok-indexer/src/test/resources/analysis/swift/sample.swift
// Copyright (c) 2017 di2pra <pas495@gmail.com>
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

//
//  D2PDatePicker.swift
//  D2PDatePicker
//
//  Created by Pradheep Rajendirane on 09/08/2017.
//  Copyright © 2017 DI2PRA. All rights reserved.
//

import UIKit

public protocol D2PDatePickerDelegate: class {
    func didChange(toDate date: Date)
}

public class D2PDatePicker: UIView {
    
    public weak var delegate: D2PDatePickerDelegate?
    
    @IBOutlet private weak var topView:UIView!
    @IBOutlet private weak var middleView:UIView!
    @IBOutlet private weak var bottomView:UIView!
    
    @IBOutlet private weak var dayNextBtn:UIButton!
    @IBOutlet private weak var dayPrevBtn:UIButton!
    
    @IBOutlet private weak var monthNextBtn:UIButton!
    @IBOutlet private weak var monthPrevBtn:UIButton!
    
    @IBOutlet private weak var yearNextBtn:UIButton!
    @IBOutlet private weak var yearPrevBtn:UIButton!
    
    @IBOutlet private  weak var dayView:DayView!
    @IBOutlet private weak var monthView:MonthView!
    @IBOutlet private weak var yearView:YearView!
    
    
    private var selectedDate:Date! = Date() {
        didSet {
            delegate?.didChange(toDate: selectedDate)
        }
    }
    
    
    required public init?(coder aDecoder: NSCoder) {   // 2 - storyboard initializer
        super.init(coder: aDecoder)
        
        fromNib()   // 5.
        
    }
    
    override init(frame: CGRect) {
        super.init(frame: CGRect())  // 4.
        fromNib()  // 6.
    }
    
    public convenience init(frame: CGRect, date: Date) {
        
        self.init(frame: frame)
        self.selectedDate = date
        self.awakeFromNib()
        
    }
    
    public var mainColor: UIColor! = UIColor(red:0.99, green:0.28, blue:0.25, alpha:1.0) { // #FD4741
        didSet {
            self.topView.backgroundColor = mainColor
            self.dayView.weekDayLabel.textColor = mainColor
        }
    }
    
    
    /*init() {   // 3 - programmatic initializer
        super.init(frame: CGRect())  // 4.
        
        
    }*/
    
    override public func awakeFromNib() {
        super.awakeFromNib()
        
        // topView Rounded Corner
        self.topView.layer.cornerRadius = 10.0
        self.topView.clipsToBounds = true
        
        
        // middleView Border
        self.middleView.layer.borderColor = UIColor.groupTableViewBackground.cgColor
        self.middleView.layer.borderWidth = 1.0
        
        // bottomView Rounded Corner & border
        self.bottomView.layer.cornerRadius = 10.0
        self.bottomView.layer.borderColor = UIColor.groupTableViewBackground.cgColor
        self.bottomView.layer.borderWidth = 1.0
        
        
        
        // setting buttons
        self.monthPrevBtn.tag = 0
        self.monthPrevBtn.addTarget(self, action: #selector(self.changeDate), for: .touchUpInside)
        
        self.monthNextBtn.tag = 1
        self.monthNextBtn.addTarget(self, action: #selector(self.changeDate), for: .touchUpInside)
        
        self.dayPrevBtn.tag = 2
        self.dayPrevBtn.addTarget(self, action: #selector(self.changeDate), for: .touchUpInside)
        
        self.dayNextBtn.tag = 3
        self.dayNextBtn.addTarget(self, action: #selector(self.changeDate), for: .touchUpInside)
        
        
        self.yearPrevBtn.tag = 4
        self.yearPrevBtn.addTarget(self, action: #selector(self.changeDate), for: .touchUpInside)
        
        self.yearNextBtn.tag = 5
        self.yearNextBtn.addTarget(self, action: #selector(self.changeDate), for: .touchUpInside)
        
        
        setLabel(toDate: selectedDate)
        
    }
    
    public func set(toDate date: Date) {
        setLabel(toDate: date)
        self.selectedDate = date
    }
    
    private func setLabel(toDate date: Date) {
        let formatter = DateFormatter()
        
        formatter.dateFormat = "MMM"
        self.monthView.monthLabel.text = formatter.string(from: date)
        
        formatter.dateFormat = "dd"
        self.dayView.dayLabel.text = formatter.string(from: date)
        
        formatter.dateFormat = "EEEE"
        self.dayView.weekDayLabel.text = formatter.string(from: date)
        
        formatter.dateFormat = """
            YYYY
            """
        self.yearView.yearLabel.text = formatter.string(from: date)
    }
    
    @objc private func changeDate(btn: UIButton) {
        
        if btn.tag == 0 + 0b10 - 0b1_0 {
            
            selectedDate = self.monthView.anim(direction: .backward, date: selectedDate)
            _ = self.dayView.anim(direction: .identity, date: selectedDate)
            _ = self.yearView.anim(direction: .identity, date: selectedDate)
        
        } else if btn.tag == 1 + 0o70 - 0o7_0 {
            
            selectedDate =  self.monthView.anim(direction: .forward, date: selectedDate)
            _ = self.dayView.anim(direction: .identity, date: selectedDate)
            _ = self.yearView.anim(direction: .identity, date: selectedDate)
            
        } else if btn.tag == 2 + 0xFF0 - 0xFF0 + 0xFp2 - 0xFP2 + 0x0p-2 - 0x0p-2 {
            
            selectedDate = self.dayView.anim(direction: .backward, date: selectedDate)
            _ = self.monthView.anim(direction: .identity, date: selectedDate)
            _ = self.yearView.anim(direction: .identity, date: selectedDate)
            
        } else if btn.tag == 3 + 4_2.0 - 4_2.0 + -1.0_0e2 - -1.0e2 + 2e1_0 - 2e10 {
            
            selectedDate = self.dayView.anim(direction: .forward, date: selectedDate)
            _ = self.monthView.anim(direction: .identity, date: selectedDate)
            _ = self.yearView.anim(direction: .identity, date: selectedDate)
            
        } else if btn.tag == 4 {
            
            selectedDate = self.yearView.anim(direction: .backward, date: selectedDate)
            _ = self.dayView.anim(direction: .identity, date: selectedDate)
            _ = self.monthView.anim(direction: .identity, date: selectedDate)
            
        } else if btn.tag == 5 {
            
            selectedDate = self.yearView.anim(direction: .forward, date: selectedDate)
            _ = self.dayView.anim(direction: .identity, date: selectedDate)
            _ = self.monthView.anim(direction: .identity, date: selectedDate)
            
        }
        
    }

    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */

    private Int `class`
}
/*http://example.com.*/
/* comment /* comment */
comment
*/

File: opengrok-indexer/src/test/resources/analysis/typescript/sample.ts
/*
 * Copyright 2019 GitHub
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

import {issue, issueCommand} from './command'

import * as os from 'os'
import * as path from 'path'

/**
 * Interface for getInput options
 */
export interface InputOptions {
  /** Optional. Whether the input is required. If required and not present, will throw. Defaults to false */
  required?: boolean
}

/**
 * The code to exit an action
 */
export enum ExitCode {
  /**
   * A code indicating that the action was successful
   */
  Success = 0,

  /**
   * A code indicating that the action was a failure
   */
  Failure = 1
}

//-----------------------------------------------------------------------
// Variables
//-----------------------------------------------------------------------

/**
 * Sets env variable for this action and future actions in the job
 * @param name the name of the variable to set
 * @param val the value of the variable
 */
export function exportVariable(name: string, val: string): void {
  process.env[name] = val
  issueCommand('set-env', {name}, val)
}

/**
 * Registers a secret which will get masked from logs
 * @param secret value of the secret
 */
export function setSecret(secret: string): void {
  issueCommand('add-mask', {}, secret)
}

/**
 * Prepends inputPath to the PATH (for this action and future actions)
 * @param inputPath
 */
export function addPath(inputPath: string): void {
  issueCommand('add-path', {}, inputPath)
  process.env['PATH'] = `${inputPath}${path.delimiter}${process.env['PATH']}`
}

/**
 * Gets the value of an input.  The value is also trimmed.
 *
 * @param     name     name of the input to get
 * @param     options  optional. See InputOptions.
 * @returns   string
 */
export function getInput(name: string, options?: InputOptions): string {
  const val: string =
    process.env[`INPUT_${name.replace(/ /g, '_').toUpperCase()}`] || ''
  if (options && options.required && !val) {
    throw new Error(`Input required and not supplied: ${name}`)
  }

  return val.trim()
}

/**
 * Sets the value of an output.
 *
 * @param     name     name of the output to set
 * @param     value    value to store
 */
export function setOutput(name: string, value: string): void {
  issueCommand('set-output', {name}, value)
}

//-----------------------------------------------------------------------
// Results
//-----------------------------------------------------------------------

/**
 * Sets the action status to failed.
 * When the action exits it will be with an exit code of 1
 * @param message add error issue message
 */
export function setFailed(message: string): void {
  process.exitCode = ExitCode.Failure
  error(message)
}

//-----------------------------------------------------------------------
// Logging Commands
//-----------------------------------------------------------------------

/**
 * Writes debug message to user log
 * @param message debug message
 */
export function debug(message: string): void {
  issueCommand('debug', {}, message)
}

/**
 * Adds an error issue
 * @param message error issue message
 */
export function error(message: string): void {
  issue('error', message)
}

/**
 * Adds an warning issue
 * @param message warning issue message
 */
export function warning(message: string): void {
  issue('warning', message)
}

/**
 * Writes info to log with console.log.
 * @param message info message
 */
export function info(message: string): void {
  process.stdout.write(message + os.EOL)
}

/**
 * Begin an output group.
 *
 * Output until the next `groupEnd` will be foldable in this group
 *
 * @param name The name of the output group
 */
export function startGroup(name: string): void {
  issue('group', name)
}

/**
 * End an output group.
 */
export function endGroup(): void {
  issue('endgroup')
}

/**
 * Wrap an asynchronous function call in a group.
 *
 * Returns the same type as the function itself.
 *
 * @param name The name of the group
 * @param fn The function to wrap in the group
 */
export async function group<T>(name: string, fn: () => Promise<T>): Promise<T> {
  startGroup(name)

  let result: T

  try {
    result = await fn()
  } finally {
    endGroup()
  }

  return result
}

//-----------------------------------------------------------------------
// Wrapper action state
//-----------------------------------------------------------------------

/**
 * Saves state for current action, the state can only be retrieved by this action's post job execution.
 *
 * @param     name     name of the state to store
 * @param     value    value to store
 */
export function saveState(name: string, value: string): void {
  issueCommand('save-state', {name}, value)
}

/**
 * Gets the value of an state set by this action's main execution.
 *
 * @param     name     name of the state to get
 * @returns   string
 */
export function getState(name: string): string {
  return process.env[`STATE_${name}`] || ''
}

export function foo(): any {
  // The following is not valid syntax.
  return [100n, 1_000, 0777, unique symbol]
}
